### PR TITLE
[ryamgoh] iP

### PR DESCRIPTION
**Pull request by ryamgoh**
# Duke

## Description

This iP PR aims to enhance our project by implementing several key features and addressing important issues. The following lists outline the changes and tasks to be completed:

### Features Added:

- [x] 🌟 Added ability to add and delete tasks
- [x] 📊 Added validation
- [x] 🛡️ Added list functionality
- [x] 🔄 Integrated persistent storage

### Issues Resolved:

1. 🐛 Fixed a critical bug causing data to fail to persist
2. 🐞 Addressed performance issues related to database queries.

### Code Changes:

```java
// Example Java code snippet with syntax highlighting
package duke;

import duke.commands.Command;
import duke.exceptions.DukeException;
import duke.parser.Parser;
import duke.storage.Storage;
import duke.tasks.TaskList;
import duke.ui.Main;
import duke.ui.Ui;
import javafx.application.Application;

public class Duke {
    /**
     * The task list
     */
    private final TaskList taskList;
    /**
     * The ui object
     */
    private final Ui ui;
    /**
     * The storage object
     */
    private final Storage storage;

    /**
     * Constructor for Duke
     *
     * @param filePath the file path
     */
    public Duke(String filePath) {
        this.storage = new Storage(filePath);
        this.taskList = new TaskList(this.storage.readData());
        this.ui = new Ui();
    }

    /**
     * Constructor for Duke
     */
    public Duke() {
        this.storage = new Storage("data/duke.txt");
        this.taskList = new TaskList(this.storage.readData());
        this.ui = new Ui();
    }

    /**
     * Initialises the program
     *
     * @param args Input args.
     */
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }

    /**
     * You should have your own function to generate a response to user input.
     * Replace this stub with your completed method.
     */
    public String getResponse(String input) {
        try {
            Command command = Parser.parse(input);
            return command.execute(this.taskList, this.ui, this.storage);
        } catch (DukeException exc) {
            return exc.getMessage();
        }
    }
}
```

### Task List:
- **Conduct a security audit to identify potential vulnerabilities.**
- **Update documentation with the latest changes.**
- **Prepare a release candidate for deployment.**

### Additional Notes:

> 
For more details, please refer to the [full PR here](https://github.com/ryamgoh/ip).

### CODE REVIEW

File: src/main/java/duke/commands/DeleteCommand.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package duke.commands;
+
+import duke.exceptions.DukeException;
+import duke.storage.Storage;
+import duke.tasks.Task;
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
+public class DeleteCommand extends Command {
```
Comment (teozern1) [inline src/main/java/duke/commands/DeleteCommand.java:12]: Good that all the commands have the same naming convention

File: src/main/java/duke/storage/Storage.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		// File doesn't exist
+		if (!this.fileExists()) {
+			this.createFile();
+		}
+	}
+
+	/**
+	 * Creates the directory
+	 */
+	public void createDirectory() {
```
Comment (teozern1) [inline src/main/java/duke/storage/Storage.java]: Good naming of methods using verbs

File: src/main/java/duke/storage/Storage.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		}
+	}
+
+	/**
+	 * Formats the string to a task
+	 *
+	 * @param line the string to be formatted
+	 * @return the task
+	 */
+	public Task formatStringToTask(String line) {
```
Comment (teozern1) [inline src/main/java/duke/storage/Storage.java]: Method name is concise but descriptive enough 👍

File: src/main/java/duke/tasks/Deadline.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+	/**
+	 * Returns the time of the deadline
+	 *
+	 * @return the time of the deadline
+	 */
+	@Override
+	public String toString() {
+		return "[D]" + super.toString()
+						+ " (by: " + this.deadline.format(DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a")) + ")";
```
Comment (teozern1) [inline src/main/java/duke/tasks/Deadline.java]: Not sure whether this is the convention for indentation?

File: src/main/java/duke/tasks/TaskList.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    public void delete(int index) {
+        this.list.remove(index);
+    }
+
+    /**
+     * Returns the size of the list
+     *
+     * @return the size of the list
+     */
+    public int length() {
```
Comment (teozern1) [inline src/main/java/duke/tasks/TaskList.java]: Should this be getLength?

**Review by teozern1 — COMMENTED** [review]
I like the use of enums for parsing the commands. The code is clean with no major issues that I can spot!

File: src/main/java/duke/parser/Parser.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
+package duke.parser;
+
+import duke.commands.*;
```
Comment (lambraydon) [inline src/main/java/duke/parser/Parser.java]: Might it be better to explicitly list each import to adhere to the coding convention?

File: src/main/java/duke/storage/Storage.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+				} else {
+					task = new Event(description, LocalDateTime.parse(interval[0], formatter), LocalDateTime.parse(interval[1], formatter), status);
+				}
+				break;
+			default:
+				task = null;
+		}
+
+		return task;
+	}
```
Comment (lambraydon) [inline src/main/java/duke/storage/Storage.java]: Not sure whether this adheres to the coding convention of switch statements?

File: src/main/java/duke/tasks/Task.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        this.description = description;
+        this.isDone = status.equals("1");
+    }
+
+    /**
+     * Returns the description of the task.
+     *
+     * @return String description of the task.
+     */
+    public String getStatusIcon() {
```
Comment (lambraydon) [inline src/main/java/duke/tasks/Task.java:44]: Good naming of methods in verb form!

**Review by lambraydon — COMMENTED** [review]
Overall LGTM! Just a few nitpicks in codestyle

**Pull request closed without merging**